nomos-membership/src/membership.rs

```diff
@@ -0,0 +1,523 @@
+                    let promoted_session = forming_session.clone();
+
+                    // Get snapshot before promoting
+                    completed_sessions.insert(*service_type, promoted_session.clone());
+
+                    // Persist the new active session
+                    self.storage
+                        .save_active_session(
+                            *service_type,
+                            promoted_session.session_number,
+                            &promoted_session.providers,
+                        )
+                        .await
+                        .map_err(MembershipError::Other)?;
+
+                    // Promote forming to active
+                    self.active_sessions
+                        .insert(*service_type, promoted_session.clone());
+
+                    // Start new forming session with incremented session number
+                    let mut new_forming = promoted_session;
+                    new_forming.session_number = next_session + 1;
+
+                    // Persist new forming session
+                    self.storage
+                        .save_forming_session(
+                            *service_type,
+                            new_forming.session_number,
+                            &new_forming.providers,
+                        )
+                        .await
+                        .map_err(MembershipError::Other)?;
+
+                    self.forming_sessions.insert(*service_type, new_forming);
+                } else {
+                    // Just persist the updated forming session
+                    self.storage
+                        .save_forming_session(
+                            *service_type,
+                            forming_session.session_number,
+                            &forming_session.providers,
+                        )
+                        .await
+                        .map_err(MembershipError::Other)?;
+                }
+            }
+        }
+
+        Ok(if completed_sessions.is_empty() {
+            None
+        } else {
+            Some(completed_sessions)
+        })
+    }
+
+    fn get_latest_providers(&self, service_type: ServiceType) -> Result<Session, MembershipError> {
+        if let Some(session_state) = self.active_sessions.get(&service_type).cloned() {
+            return Ok(session_state);
+        }
+
+        Err(MembershipError::NotFound)
+    }
+}
+
+impl Session {
+    fn apply_update(&mut self, update: &FinalizedBlockEventUpdate) {
+        match update.state {
+            nomos_core::sdp::FinalizedDeclarationState::Active => {
+                self.providers
+                    .insert(update.provider_id, update.locators.clone());
+            }
+            nomos_core::sdp::FinalizedDeclarationState::Inactive
+            | nomos_core::sdp::FinalizedDeclarationState::Withdrawn => {
+                self.providers.remove(&update.provider_id);
+            }
+        }
+    }
+}
```

Comment:
Why isnt this in a session.rs module with its actual struct instead?

Reply:
Good catch, it is a leftover. Tnx!